llmfoundry/data/finetuning/tasks.py

```diff
@@ -102,6 +112,72 @@ def _is_empty_or_nonexistent(dirpath: str) -> bool:
     return not os.path.isdir(dirpath) or len(os.listdir(dirpath)) == 0
 
 
+def _get_role_key(message: Dict[str, str]) -> str:
+    if not isinstance(message, Mapping):
+        raise TypeError(
+            f'Expected message to be a mapping, but found {type(message)}')
+    role_keys = _ALLOWED_ROLE_KEYS.intersection(message.keys())
+    if len(role_keys) != 1:
+        raise ValueError(f'Expected 1 role key, but found {len(role_keys)}')
+    role_key = list(role_keys)[0]
+    return role_key
+
+
+def _get_content_key(message: Dict[str, str]) -> str:
```

Comment:
these get functions are pretty similar in their implementations, can we generalize?

Reply:
What about some kind of `safe_get` function to safely get from a mapping? This could also be used for prompt response checks

overly simplified pseudo code
```
def safe_get(dictionary: dict, allowed_keys: set, type: Any):
    ...
    if not isinstance(value, type):
        raise TypeError()
    return value


def validate_chat_formatted_data(example: dict):
    messages = safe_get(example, _ALLOWED_MESSAGES_KEYS, type=list) 
    for idx, message in enumerate(messages):
        role = safe_get(example, _ALLOWED_ROLE_KEYS, type=str)
        _ = safe_get(example, _ALLOWED_CONTENT_KEYS, type=str)
        
        if role not in _ALLOWED_ROLES:
            raise ValueError()
        
        if idx == len(messages) - 1:
            if role not in _ALLOWED_LAST_MESSAGE_ROLES:
                raise ValueError()

```